hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java

```diff
@@ -0,0 +1,200 @@
+            }
+            if (!matchingColumnFound) {
+                throw new MetaModelException(String.format("ColumnFamily: %s doesn't exist in the schema of the table",
+                        columnNameOfCheckedTable));
+            }
+        }
+    }
+
+    /**
+     * Creates a set of columnFamilies out of a list of hbaseColumns
+     *
+     * @param columns
+     * @return {@link LinkedHashSet}
+     */
+    private static Set<String> getColumnFamilies(final HBaseColumn[] columns) {
+        return Arrays.stream(columns).map(HBaseColumn::getColumnFamily).distinct().collect(Collectors.toSet());
+    }
+
+    @Override
+    public synchronized void execute() {
+        if (_indexOfIdColumn == -1) {
+            throw new MetaModelException("The ID-Column was not found");
+        }
+
+        // The columns parameter should match the table's columns, just to be sure, this is checked again
+        checkColumnFamilies((HBaseTable) getTable(), getColumnFamilies(getColumns()));
+
+        ((HBaseDataContext) getUpdateCallback().getDataContext()).getHBaseClient().insertRow(getTable().getName(),
+                getColumns(), getValues(), _indexOfIdColumn);
+    }
+
+    @Override
+    protected HBaseColumn[] getColumns() {
+        return columns.toArray(new HBaseColumn[columns.size()]);
+    }
+
+    @Override
+    protected Object[] getValues() {
+        return values.toArray(new Object[values.size()]);
+    }
+
+    @Override
+    public RowInsertionBuilder value(final Column column, final Object value, final Style style) {
+        if (column == null) {
+            throw new IllegalArgumentException("Column cannot be null.");
+        }
+
+        final HBaseColumn hbaseColumn = getHbaseColumn(column);
+
+        for (int i = 0; i < columns.size(); i++) {
+            if (columns.get(i).equals(hbaseColumn)) {
+                values.set(i, value);
+                return this;
+            }
+        }
+
+        if (hbaseColumn.isPrimaryKey()) {
+            _indexOfIdColumn = columns.size();
+        }
+
+        columns.add((HBaseColumn) hbaseColumn);
+        values.add(value);
+
+        return this;
+    }
+
+    private HBaseColumn getHbaseColumn(final Column column) {
+        if (column instanceof HBaseColumn) {
+            return (HBaseColumn) column;
+        } else {
+            final String columnName = column.getName();
+            final String[] columnNameParts = columnName.split(":");
+            if (columnNameParts.length == 1) {
+                return new HBaseColumn(columnNameParts[0], getTable());
+            }
+            if (columnNameParts.length == 2) {
+                return new HBaseColumn(columnNameParts[0], columnNameParts[1], getTable());
+            }
+            throw new MetaModelException("Can't determine column family for column \"" + columnName + "\".");
+        }
+    }
+
+    @Override
+    public RowInsertionBuilder value(final int columnIndex, final Object value) {
```

Comment:
When you extend the abstract class that you use here, you shouldn't need to override these methods. Is there some special reason for doing that? I worry that any new method addition will be forgotten here and would rather that we just used the abstract class'es methods.

Reply:
The special reason is that we don't use the `_values` and `_columns` arrays used by the super classes, because when creating the insertion builder we can only pass the Table in the constructor and the available columns on an HBase table aren't fixed, because qualifiers are dynamic can be added on the fly, only the column families are fixed.

Reply:
In that case, shouldn't we then avoid the abstract class, and instead do a full implementation of the interface?

Reply:
Wait ... So how would a user be able to pass in a column that's not already defined? I see that `HBaseTable` returns `HBaseColumn`s which represent that paths down into the column families as defined by the `SimpleTableDef`s. So it's kind of dynamic based on that `SimpleTableDef`. But is there some way that the columns will change after creating the simple table def? Or rather - in the middle of a update script? I don't see how that would be possible, but maybe it is, and then I think that deserves some clarification (since it's unusual for MetaModel).

Reply:
The columns are not defined, only the column families are, qualifiers are not. So you can insert values into columns which don't exist yet. Each row in an HBase table can have different columns from other rows in that table. Effectively that makes it harder to read from an HBase table in a manner that fits MetaModel then to write to an HBase table, because MetaModel typically writes the to same columns over and over again and that's no problem for HBase.

Reply:
I'm going to look a bit into the necessity of the using the abstract class, maybe a full implementation is indeed better in this case.

Reply:
I refactored the class so it no longer extend the `AbstractRowInsertionBuilder`, but implements its logic completely by itself.